File: Sources/CDSP/Engine/engine_capture_loop.c
// Capture thread body. One instance per engine run; the thread
// closure invokes `run()` exactly once and returns when the shared
// `shouldStop` flag is set or a stop reason is reported.
//
// State ownership
// ---------------
// All mutable state — the working chunk, the silence counter, the
// stall watchdog — lives inside the loop instance and is touched
// only by the capture thread. Cross-thread communication happens
// exclusively through the injected `EngineSharedState`.
//
// Audio-thread invariants
// -----------------------
//   * No allocations in the steady-state. Audio chunks are obtained
//     from a pre-allocated `RoundRobinChunkPool`.
//   * No locks. Coordination uses the shared SPSC queue + semaphore.
//   * No `Date()` / `gettimeofday`. The watchdog uses
//     `clock_gettime_nsec_np(CLOCK_UPTIME_RAW)` (vDSO read on
//     Darwin — no syscall).
#include "engine_capture_loop.h"

#include <stdio.h>

#include "Audio/silence_counter.h"
#include "sample_rate_watcher.h"

struct engine_capture_loop {
  engine_shared_state_t* shared;
  engine_state_machine_t* state_machine;
  capture_backend_t* capture;
  playback_backend_t* playback;
  processing_parameters_t* processing_params;
  dop_decoder_t* dop_decoder;

  size_t chunk_size;
  size_t channels;
  size_t samplerate;
  double last_observed_pending_rate;
  bool has_last_observed_pending_rate;
  double last_observed_playback_pending_rate;
  bool has_last_observed_playback_pending_rate;

  silence_counter_t* silence_counter;
  round_robin_chunk_pool_t* chunk_pool;

  uint64_t watchdog_last_success_ns;
  bool watchdog_triggered;
  double watchdog_timeout_seconds;

  sample_rate_watcher_t* rate_watcher;
};
#include <stdlib.h>
#include <time.h>

#include "Logging/app_logger.h"
#include "thread_priority.h"

#ifndef __APPLE__
#define CLOCK_UPTIME_RAW CLOCK_MONOTONIC
/**
 * @brief Helper function to retrieve the raw system uptime in nanoseconds.
 * Used for the watchdog stall detector. On non-Apple platforms, this wraps
 * clock_gettime(CLOCK_MONOTONIC).
 *
 * @param clock_id The system clock identifier.
 * @return The current time value in nanoseconds.
 */
static inline uint64_t clock_gettime_nsec_np(int clock_id) {
  struct timespec ts;
  clock_gettime(clock_id, &ts);
  return (uint64_t)ts.tv_sec * 1000000000ULL + ts.tv_nsec;
}

#endif

engine_capture_loop_t* engine_capture_loop_create(
    engine_shared_state_t* shared, engine_state_machine_t* state_machine,
    capture_backend_t* capture, playback_backend_t* playback,
    processing_parameters_t* processing_params, dop_decoder_t* dop_decoder,
    round_robin_chunk_pool_t* chunk_pool, size_t chunk_size, size_t channels,
    size_t samplerate, double silence_threshold_db,
    double silence_timeout_seconds, bool stop_on_rate_change,
    double rate_measure_interval) {
  engine_capture_loop_t* loop =
      (engine_capture_loop_t*)calloc(1, sizeof(engine_capture_loop_t));
  if (!loop) return NULL;

  loop->shared = shared;
  loop->state_machine = state_machine;
  loop->capture = capture;
  loop->playback = playback;
  loop->processing_params = processing_params;
  loop->dop_decoder = dop_decoder;
  loop->chunk_pool = chunk_pool;
  loop->chunk_size = chunk_size;
  loop->channels = channels;
  loop->samplerate = samplerate;
  loop->silence_counter = silence_counter_create(
      silence_threshold_db, silence_timeout_seconds, samplerate, chunk_size);
  if (!loop->silence_counter) {
    engine_capture_loop_free(loop);
    return NULL;
  }

  loop->rate_watcher = sample_rate_watcher_create(
      (double)samplerate, rate_measure_interval, stop_on_rate_change);
  if (!loop->rate_watcher) {
    engine_capture_loop_free(loop);
    return NULL;
  }

  loop->watchdog_timeout_seconds = 0.5;
  loop->watchdog_last_success_ns = clock_gettime_nsec_np(CLOCK_UPTIME_RAW);
  loop->watchdog_triggered = false;

  return loop;
}

void engine_capture_loop_free(engine_capture_loop_t* loop) {
  if (!loop) return;
  if (loop->silence_counter) {
    silence_counter_free(loop->silence_counter);
  }
  if (loop->rate_watcher) {
    sample_rate_watcher_free(loop->rate_watcher);
  }
  free(loop);
}

void engine_capture_loop_run(engine_capture_loop_t* loop) {
  if (!loop) return;
  logger_t logger = logger_create("dsp.capture");
  logger_info(&logger, "Capture thread started", log_arg_none(), log_arg_none(),
              log_arg_none(), log_arg_none());

  // Set real-time execution priority parameters for this thread.
  set_realtime_thread_priority("Capture", loop->chunk_size, loop->samplerate);

  sample_rate_watcher_reset(loop->rate_watcher);

  while (
      !atomic_load_explicit(&loop->shared->should_stop, memory_order_acquire)) {
    if (engine_state_machine_get_state(loop->state_machine) ==
        PROCESSING_STATE_PAUSED) {
      sample_rate_watcher_reset(loop->rate_watcher);
    }
    // 1. Hardware Sample-Rate Change Check:
    // Check if the hardware sample rates have drifted or been explicitly
    // modified (e.g. by another application or OS settings). An unexpected
    // hardware rate change invalidates the processing thread pipeline, so we
    // signal a host rebuild stop reason.
    double rate = 0.0;
    if (capture_backend_get_pending_rate_change(loop->capture, &rate)) {
      if (!loop->has_last_observed_pending_rate ||
          rate != loop->last_observed_pending_rate) {
        loop->last_observed_pending_rate = rate;
        loop->has_last_observed_pending_rate = true;
        logger_warn(&logger,
                    "Capture device rate changed to %f Hz; stopping engine",
                    log_arg_double(rate), log_arg_none(), log_arg_none(),
                    log_arg_none());
        processing_stop_reason_t reason = {
            .type = STOP_REASON_CAPTURE_FORMAT_CHANGE,
            .format_change_rate = (int)(rate + 0.5)};
        engine_shared_state_request_stop(loop->shared, reason);
        break;
      }
    }
    if (playback_backend_get_pending_rate_change(loop->playback, &rate)) {
      if (!loop->has_last_observed_playback_pending_rate ||
          rate != loop->last_observed_playback_pending_rate) {
        loop->last_observed_playback_pending_rate = rate;
        loop->has_last_observed_playback_pending_rate = true;
        logger_warn(&logger,
                    "Playback device rate changed to %f Hz; stopping engine",
                    log_arg_double(rate), log_arg_none(), log_arg_none(),
                    log_arg_none());
        processing_stop_reason_t reason = {
            .type = STOP_REASON_PLAYBACK_FORMAT_CHANGE,
            .format_change_rate = (int)(rate + 0.5)};
        engine_shared_state_request_stop(loop->shared, reason);
        break;
      }
    }

    // 2. Fetch a chunk buffer from the pre-allocated round-robin pool.
    audio_chunk_t* chunk = round_robin_chunk_pool_next(loop->chunk_pool);
    backend_error_t err;
    backend_error_init(&err, BACKEND_ERROR_NONE, "");

    // 3. Read raw PCM/DSD frame data from the capture backend.
    bool got_data =
        capture_backend_read(loop->capture, loop->chunk_size, chunk, &err);
    if (!got_data) {
      if (err.type == BACKEND_ERROR_READ_EOF) {
        logger_info(&logger,
                    "Capture reached End-of-Stream; stopping engine gracefully",
                    log_arg_none(), log_arg_none(), log_arg_none(),
                    log_arg_none());
        processing_stop_reason_t reason = {.type = STOP_REASON_DONE};
        snprintf(reason.message, sizeof(reason.message), "EOF");
        engine_shared_state_request_stop(loop->shared, reason);
        break;
      }
      // If reading fails with an error, trigger an engine stop.
      if (err.type != BACKEND_ERROR_NONE) {
        logger_error(&logger, "Capture error: %s",
                      log_arg_string(err.message), log_arg_none(),
                      log_arg_none(), log_arg_none());
        processing_stop_reason_t reason = {.type = STOP_REASON_CAPTURE_ERROR};
        snprintf(reason.message, sizeof(reason.message), "%s", err.message);
        engine_shared_state_request_stop(loop->shared, reason);
        break;
      }
      // If we should stop, exit the thread loop.
      if (atomic_load_explicit(&loop->shared->should_stop,
                               memory_order_acquire))
        break;
      // If the engine is in a PAUSED state (no active input signal), reset the
      // watchdog timer to avoid triggering stall warnings while waiting for
      // signal.
      if (engine_state_machine_get_state(loop->state_machine) ==
          PROCESSING_STATE_PAUSED) {
        loop->watchdog_last_success_ns =
            clock_gettime_nsec_np(CLOCK_UPTIME_RAW);
        capture_backend_wait(loop->capture, 20);
        continue;
      }
      // 4. Watchdog / Stall Monitor:
      // If the engine is running but we get no data chunks from the capture
      // device for more than watchdog_timeout_seconds, set state to STALLED and
      // log a warning.
      if (!loop->watchdog_triggered) {
        uint64_t now = clock_gettime_nsec_np(CLOCK_UPTIME_RAW);
        double elapsed =
            (double)(now - loop->watchdog_last_success_ns) / 1000000000.0;
        if (elapsed > loop->watchdog_timeout_seconds) {
          loop->watchdog_triggered = true;
          engine_state_machine_set_state(loop->state_machine,
                                         PROCESSING_STATE_STALLED);
          logger_warn(&logger, "Capture device stalled — no data for %fs",
                      log_arg_double(loop->watchdog_timeout_seconds),
                      log_arg_none(), log_arg_none(), log_arg_none());
        }
      }
      // Block/wait up to 20ms using the backend's synchronization mechanism
      // (e.g. semaphore). This yields CPU time while maintaining real-time
      // scheduling priority.
      capture_backend_wait(loop->capture, 20);
      continue;
    }

    // 5. Watchdog Stall Recovery:
    // Reset the watchdog status if we successfully read data after a stall.
    loop->watchdog_last_success_ns = clock_gettime_nsec_np(CLOCK_UPTIME_RAW);
    if (loop->watchdog_triggered) {
      loop->watchdog_triggered = false;
      logger_info(&logger, "Capture recovered from stall", log_arg_none(),
                  log_arg_none(), log_arg_none(), log_arg_none());
    }

    // 5.5. Rate Watcher Measurement:
    double measured_rate = 0.0;
    if (sample_rate_watcher_tick(loop->rate_watcher, loop->chunk_size,
                                 &measured_rate)) {
      logger_warn(&logger,
                  "Sample rate change detected (measured: %f Hz, expected: %zu "
                  "Hz); stopping engine",
                  log_arg_double(measured_rate), log_arg_int(loop->samplerate),
                  log_arg_none(), log_arg_none());
      if (sample_rate_watcher_get_stop_on_rate_change(loop->rate_watcher)) {
        processing_stop_reason_t reason = {
            .type = STOP_REASON_CAPTURE_FORMAT_CHANGE,
            .format_change_rate = (int)(measured_rate + 0.5)};
        engine_shared_state_request_stop(loop->shared, reason);
        break;
      }
    }

    // 6. DoP (DSD over PCM) Decoding:
    // If DoP decoding is active, process the PCM chunk to detect DSD marker
    // flags and decode them back to raw DSD samples in-place. Decoding is done
    // before metering so RMS/Peak values reflect the actual signal instead of
    // the carrier noise.
    if (loop->dop_decoder) {
      dop_decoder_detect_and_process(loop->dop_decoder, chunk);
    }

    // Update level meters with the peak/rms of this chunk.
    double loudest_peak = processing_parameters_update_capture_levels(
        loop->processing_params, chunk);

    // 7. Silence/Auto-pause Gate:
    // Update the silence counter. If the signal level is below the threshold
    // for longer than the timeout duration, desired is set to
    // PROCESSING_STATE_PAUSED. We toggle the backends' state to paused to stop
    // downstream devices.
    processing_state_t desired =
        silence_counter_update(loop->silence_counter, loudest_peak);
    processing_state_t current =
        engine_state_machine_get_state(loop->state_machine);
    if (desired != current) {
      engine_state_machine_set_state(loop->state_machine, desired);
      playback_backend_set_is_paused(loop->playback,
                                     (desired == PROCESSING_STATE_PAUSED));
      capture_backend_set_is_paused(loop->capture,
                                    (desired == PROCESSING_STATE_PAUSED));
    }

    // 8. Enqueue Captured Chunk:
    // If the engine is running (not paused), push the chunk pointer into the
    // bounded lock-free SPSC queue. If the queue is full, sleep briefly to
    // yield CPU and propagate backpressure upstream.
    if (engine_state_machine_get_state(loop->state_machine) !=
        PROCESSING_STATE_PAUSED) {
      while (!spsc_queue_enqueue(loop->shared->captured_queue, chunk)) {
        if (atomic_load_explicit(&loop->shared->should_stop,
                                 memory_order_acquire) &&
            loop->shared->stop_reason.type != STOP_REASON_DONE) {
          break;
        }
        engine_yield();
      }
      // Signal the processing thread that a new chunk is available.
      engine_sem_signal(loop->shared->captured_semaphore);
    }
  }

  // Propagate graceful shutdown sequentially.
  atomic_store_explicit(&loop->shared->capture_finished, true,
                        memory_order_release);
  engine_sem_signal(loop->shared->captured_semaphore);

  logger_info(&logger, "Capture thread stopped", log_arg_none(), log_arg_none(),
              log_arg_none(), log_arg_none());
}

File: Sources/SwiftDSP/Engine/EngineCaptureLoop.swift
// Capture thread body. One instance per engine run; the thread
// closure invokes `run()` exactly once and returns when the shared
// `shouldStop` flag is set or a stop reason is reported.
//
// State ownership
// ---------------
// All mutable state — the working chunk, the silence counter, the
// stall watchdog — lives inside the loop instance and is touched
// only by the capture thread. Cross-thread communication happens
// exclusively through the injected `EngineSharedState`.
//
// Audio-thread invariants
// -----------------------
//   * No allocations in the steady-state. Audio chunks are obtained
//     from a pre-allocated `RoundRobinChunkPool`.
//   * No locks. Coordination uses the shared SPSC queue + semaphore.
//   * No `Date()` / `gettimeofday`. The watchdog uses
//     `clock_gettime_nsec_np(CLOCK_UPTIME_RAW)` (vDSO read on
//     Darwin — no syscall).

import DSPConfig
import Foundation
import Synchronization

/// `@unchecked Sendable` is a *transfer* vouch, not a *share*
/// vouch: the instance is safe to cross the Thread spawn boundary
/// because exactly one thread (the loop thread) ever touches it
/// after `run()` is invoked. The mutable state — the working
/// `AudioChunk`, the silence counter, the stall watchdog — has no
/// internal synchronisation and is *not* safe to use from multiple
/// threads concurrently.
final class EngineCaptureLoop: @unchecked Sendable {
  private let logger = Logger(label: "dsp.capture")

  private let shared: EngineSharedState
  private let stateMachine: EngineStateMachine
  private let capture: CaptureBackend
  private let playback: PlaybackBackend
  private let processingParams: ProcessingParameters
  private var dopDecoder: DoPDecoder

  private let chunkSize: Int
  private let channels: Int
  private let samplerate: Int
  private var lastObservedPendingRate: Double?
  private var lastObservedPlaybackPendingRate: Double?

  /// Hooked stop callback. Invoked when capture decides the engine
  /// must shut down (format change / capture error / stall). The
  /// host wires this to `DSPEngineCore.stop(reason:)` so the once-CAS
  /// teardown runs exactly once even when several signals fire
  /// concurrently.
  private let onStop: (ProcessingStopReason) -> Void

  init(
    shared: EngineSharedState,
    stateMachine: EngineStateMachine,
    capture: CaptureBackend,
    playback: PlaybackBackend,
    processingParams: ProcessingParameters,
    dopDecoder: DoPDecoder,
    chunkSize: Int,
    channels: Int,
    samplerate: Int,
    silenceThresholdDb: Double,
    silenceTimeoutSeconds: Double,
    stopOnRateChange: Bool,
    rateMeasureInterval: Double,
    onStop: @escaping (ProcessingStopReason) -> Void
  ) {
    self.shared = shared
    self.stateMachine = stateMachine
    self.capture = capture
    self.playback = playback
    self.processingParams = processingParams
    self.dopDecoder = dopDecoder
    self.chunkSize = chunkSize
    self.channels = channels
    self.samplerate = samplerate
    self.onStop = onStop
    self.silenceCounter = SilenceCounter(
      thresholdDb: silenceThresholdDb,
      timeoutSeconds: silenceTimeoutSeconds,
      samplerate: samplerate,
      chunksize: chunkSize
    )
    self.rateWatcher = SampleRateWatcher(
      targetRate: Double(samplerate),
      measureIntervalSeconds: rateMeasureInterval,
      stopOnRateChange: stopOnRateChange
    )
  }

  // Loop-private state.
  private var silenceCounter: SilenceCounter
  private var watchdog = StallWatchdog(timeoutSeconds: 0.5)
  private var rateWatcher: SampleRateWatcher

  func run() {
    logger.info("Capture thread started")
    setRealtimeThreadPriority(name: "Capture", bufferFrames: chunkSize, sampleRate: samplerate)

    rateWatcher.reset()
    var chunkPool = RoundRobinChunkPool(
      capacity: shared.capturedQueue.capacity + 4,
      frames: chunkSize,
      channels: channels
    )

    while !shared.shouldStop.load(ordering: .acquiring) {
      if stateMachine.state == .paused {
        rateWatcher.reset()
      }

      // Surface a HAL-level sample-rate change before doing any
      // more work. A user (or another app) flipping the device
      // rate in Audio MIDI Setup invalidates the AudioUnit's
      // configured format; the cleanest recovery is to stop
      // unconditionally and let the host rebuild.
      if let rate = capture.pendingRateChange {
        if rate != lastObservedPendingRate {
          lastObservedPendingRate = rate
          logger.warning("Capture device rate changed to %f Hz; stopping engine", .double(rate))
          onStop(.captureFormatChange(Int(rate.rounded())))
          return
        }
      }
      if let rate = playback.pendingRateChange {
        if rate != lastObservedPlaybackPendingRate {
          lastObservedPlaybackPendingRate = rate
          logger.warning("Playback device rate changed to %f Hz; stopping engine", .double(rate))
          onStop(.playbackFormatChange(Int(rate.rounded())))
          return
        }
      }

      do {
        var chunk = chunkPool.next()
        let gotData = try capture.read(frames: chunkSize, into: &chunk)
        if !gotData {
          handleEmptyRead()
          continue
        }
        watchdog.onSuccessfulRead { logger.info("Capture recovered from stall") }

        // Increment captured frames and check sample rate drift.
        if let measuredRate = rateWatcher.tick(frames: chunkSize) {
          if rateWatcher.stopOnRateChange {
            logger.warning(
              "Sample rate change detected (measured: %f Hz, expected: %d Hz); stopping engine",
              .double(measuredRate), .int(samplerate)
            )
            onStop(.captureFormatChange(Int(measuredRate.rounded())))
            return
          } else {
            logger.warning(
              "Sample rate change detected (measured: %f Hz, expected: %d Hz); stop on rate change is disabled, continuing",
              .double(measuredRate), .int(samplerate)
            )
          }
        }

        // Decode DoP in place before computing capture levels so the
        // monitoring meters reflect the actual decoded audio rather
        // than the carrier waveform with its high-frequency marker
        // bytes (which would otherwise show a tiny ~0.04 amplitude
        // floor).
        do {
          _ = try dopDecoder.detectAndProcess(chunk: &chunk)
        } catch {
          logger.error("DoP decode error: %s", .string("\(error)"))
        }

        let loudestPeak = processingParams.updateCaptureLevels(from: chunk)

        // Update silence detector with the loudest channel's peak.
        // We only flip when the value actually changes to avoid
        // hammering the atomic from the audio thread.
        let desired = silenceCounter.update(signalPeakDb: loudestPeak)
        let current = stateMachine.state
        if desired != current {
          stateMachine.setState(desired)
          playback.isPaused = (desired == .paused)
        }

        // Enqueue for processing. The lock-free SPSC queue is bounded.
        // On overflow (processing loop is slow), we block the capture thread
        // to yield CPU and propagate backpressure upstream.
        if stateMachine.state != .paused {
          while !shared.capturedQueue.enqueue(chunk) {
            if shared.shouldStop.load(ordering: .acquiring) && stateMachine.stopReason != .done {
              break
            }
            Thread.sleep(forTimeInterval: 0.002)
          }
          shared.capturedSemaphore.signal()
        }
      } catch {
        logger.error("Capture error: %s", .string("\(error)"))
        onStop(.captureError("\(error)"))
        return
      }
    }
    shared.captureFinished.store(true, ordering: .releasing)
    shared.capturedSemaphore.signal()

    logger.info("Capture thread stopped")
  }

  private func handleEmptyRead() {
    if shared.shouldStop.load(ordering: .acquiring) { return }
    if stateMachine.state == .paused {
      watchdog.reset()
      _ = capture.wait(timeout: .now() + .milliseconds(20))
      return
    }
    if watchdog.tickEmptyRead() {
      stateMachine.setState(.stalled)
      logger.warning("Capture device stalled — no data for %fs", .double(watchdog.timeoutSeconds))
    }

    // Wait on the capture device's GCD semaphore for new samples, up to 20ms.
    // This uses a 20ms timeout design, preserving
    // real-time priority propagation under load instead of doing a raw sleep.
    _ = capture.wait(timeout: .now() + .milliseconds(20))
  }
}

// MARK: - SilenceCounter

/// Counts consecutive silent chunks against a dB threshold and
/// reports back the desired engine state. `update(signalPeakDb:)`
/// returns `.paused` once silence has persisted for at least the
/// configured timeout, `.running` otherwise.
///
/// Disabled when `timeoutSeconds <= 0` — in that case `update`
/// always returns `.running`.
struct SilenceCounter {
  private let limitChunks: Int
  private let thresholdDb: Double
  private var silentChunks: Int = 0

  init(thresholdDb: Double, timeoutSeconds: Double, samplerate: Int, chunksize: Int) {
    self.thresholdDb = thresholdDb
    if timeoutSeconds > 0, chunksize > 0 {
      self.limitChunks = Int((timeoutSeconds * Double(samplerate) / Double(chunksize)).rounded())
    } else {
      self.limitChunks = 0
    }
  }

  /// Feed the next chunk's loudest channel peak (dB). Returns the
  /// engine state the capture loop should drive to.
  mutating func update(signalPeakDb: Double) -> ProcessingState {
    guard limitChunks > 0 else { return .running }
    if signalPeakDb > thresholdDb {
      silentChunks = 0
      return .running
    }
    if silentChunks < limitChunks {
      silentChunks += 1
    }
    return silentChunks >= limitChunks ? .paused : .running
  }
}

// MARK: - StallWatchdog

/// Detects a hung capture device — `read` returning no data for
/// longer than `timeoutSeconds` consecutively. The watchdog records
/// the monotonic time of the most recent successful read and reports
/// `true` exactly once per stall (subsequent ticks return `false`
/// until the next successful read clears the flag).
///
/// Backed by `clock_gettime_nsec_np(CLOCK_UPTIME_RAW)` — a vDSO
/// read on Darwin, no syscall, suitable for invocation on every
/// audio-thread iteration.
struct StallWatchdog {
  let timeoutSeconds: Double
  private var lastSuccessNs: UInt64
  private var triggered: Bool = false

  init(timeoutSeconds: Double) {
    self.timeoutSeconds = timeoutSeconds
    self.lastSuccessNs = clock_gettime_nsec_np(CLOCK_UPTIME_RAW)
  }

  /// Called when `capture.read` returns no data. Returns `true` the
  /// first time the empty-read window crosses `timeoutSeconds`; a
  /// repeated call before the next successful read returns `false`.
  mutating func tickEmptyRead() -> Bool {
    if triggered { return false }
    let now = clock_gettime_nsec_np(CLOCK_UPTIME_RAW)
    let elapsed = Double(now &- lastSuccessNs) / 1_000_000_000.0
    if elapsed > timeoutSeconds {
      triggered = true
      return true
    }
    return false
  }

  /// Called after a successful read. If the watchdog had previously
  /// fired, invoke `onRecovery` once before clearing the flag.
  mutating func onSuccessfulRead(_ onRecovery: () -> Void) {
    lastSuccessNs = clock_gettime_nsec_np(CLOCK_UPTIME_RAW)
    if triggered {
      triggered = false
      onRecovery()
    }
  }

  /// Reset the watchdog last success timestamp and clear the triggered flag.
  mutating func reset() {
    lastSuccessNs = clock_gettime_nsec_np(CLOCK_UPTIME_RAW)
    triggered = false
  }
}

// MARK: - SampleRateWatcher

/// Monitors the actual capture rate by counting incoming sample frames
/// over a rolling time window. If the measured rate deviates significantly
/// from the expected target rate, it flags a rate change.
struct SampleRateWatcher {
  let stopOnRateChange: Bool
  private let targetRate: Double
  private let measureIntervalSeconds: Double
  private let tolerance: Double = 0.04
  private let consecLimit: Int = 3

  private var capturedFrames: Int = 0
  private var lastResetNs: UInt64 = 0
  private var deviationCount: Int = 0

  init(targetRate: Double, measureIntervalSeconds: Double, stopOnRateChange: Bool) {
    self.targetRate = targetRate
    self.measureIntervalSeconds = measureIntervalSeconds
    self.stopOnRateChange = stopOnRateChange
  }

  /// Reset the rate watcher stats (e.g. when unpausing or starting).
  mutating func reset() {
    capturedFrames = 0
    lastResetNs = clock_gettime_nsec_np(CLOCK_UPTIME_RAW)
    deviationCount = 0
  }

  /// Record a successfully read chunk of size `frames`.
  /// Returns `measuredRate` (Double) if the measurement interval expired and a rate change was detected,
  /// otherwise returns `nil`.
  mutating func tick(frames: Int) -> Double? {
    if lastResetNs == 0 {
      lastResetNs = clock_gettime_nsec_np(CLOCK_UPTIME_RAW)
    }
    capturedFrames += frames
    let now = clock_gettime_nsec_np(CLOCK_UPTIME_RAW)
    let elapsed = Double(now &- lastResetNs) / 1_000_000_000.0

    guard elapsed >= measureIntervalSeconds else { return nil }

    let measuredRate = Double(capturedFrames) / elapsed
    capturedFrames = 0
    lastResetNs = now

    let minVal = targetRate / (1.0 + tolerance)
    let maxVal = targetRate * (1.0 + tolerance)

    if measuredRate < minVal || measuredRate > maxVal {
      deviationCount += 1
    } else {
      deviationCount = 0
    }

    if deviationCount >= consecLimit {
      return measuredRate
    }
    return nil
  }
}
